deliver/lib/deliver/upload_screenshots.rb

```diff
@@ -52,25 +55,178 @@ def upload(options, screenshots, max_n_threads = 16)
+      end
+
+      # then, compare the new screenshots with the existing ones
+      sets_per_language.each do |language, screenshots_per_device_type|
+        updated_screenshots_per_device_type = {}
+        localization = localizations.find do |l|
+          l.locale == language
+        end
+
+        unless localization
+          UI.error("Couldn't find localization on version for #{language}")
+          next
+        end
+
+        app_screenshot_sets_map = {}
+        app_screenshot_sets = localization.get_app_screenshot_sets
+        app_screenshot_sets.each do |app_screenshot_set|
+          app_screenshot_sets_map[app_screenshot_set.screenshot_display_type] = app_screenshot_set
+        end
+
+        screenshots_per_device_type.each do |device_type, screenshots_with_checksums|
+          existing_screenshots = app_screenshot_sets_map[device_type].app_screenshots
+          updated_screenshots_per_device_type[device_type] ||= []
+
+          temp_checksums = []
+          screenshots_with_checksums.each_with_index do |screenshot_with_checksum, index|
+            if index >= 10
+              UI.error("Too many screenshots found for device '#{device_type}' in '#{language}', skipping this one (#{screenshot_with_checksum[:screenshot].path})")
+              next
+            end
+
+            # skip over duplicates
+            if temp_checksums.include?(screenshot_with_checksum[:checksum])
+              UI.message("Duplicate found. Skipping '#{screenshot.path}'...")
+              next
+            end
+            temp_checksums << screenshot_with_checksum[:checksum]
+
+            existing_screenshot = existing_screenshots[index]
+            next unless existing_screenshot.nil? || existing_screenshot.source_file_checksum != screenshot_with_checksum[:checksum]
+
+            updated_screenshots_per_device_type[device_type] << {
+                screenshot: screenshot_with_checksum[:screenshot],
+                position: index
+            }
+          end
+        end
+
+        updated_sets_per_language[language] = updated_screenshots_per_device_type
+
+        count = updated_screenshots_per_device_type.values.reduce(0) { |sum, screenshots_with_positions| sum + screenshots_with_positions.size }
+        UI.message("Found #{count} updated screenshots for language #{language}")
+      end
+
+      updated_sets_per_language
+    end
+
+    def get_checksums_to_delete(localizations, updated_sets_per_language)
+      checksums_to_delete = {} # per language and device type
+
+      updated_sets_per_language.each do |language, updated_screenshots_per_device_type|
+        checksums_per_device_type = {}
+        localization = localizations.find do |l|
+          l.locale == language
+        end
+
+        unless localization
+          UI.error("Couldn't find localization on version for #{language}")
+          next
+        end
+
+        app_screenshot_sets = localization.get_app_screenshot_sets
```

Comment:
Note: this is an expensive operation (requires a network call to Apple since it's **not** cached). I'd suggest re-architecting the code a bit so that the existing sets are only retrieved once and then passed to all the methods that need them.

Reply:
I extracted it as a separate operation but, unfortunately, it needs to be refreshed when you are counting the screenshots that are remaining in App Store (to check if they were really deleted) and before the upload procedure.